controller/node_controller.go

```diff
@@ -1018,3 +1003,250 @@ func BackingImageDiskFileCleanup(node *longhorn.Node, bi *longhorn.BackingImage,
 		delete(bi.Spec.Disks, diskUUID)
 	}
 }
+
+func (nc *NodeController) checkMonitor(nodeName string) (monitor.Monitor, error) {
+	if nodeName == "" {
+		return nil, nil
+	}
+
+	if nc.monitor != nil {
+		return nc.monitor, nil
+	}
+
+	monitor, err := monitor.NewNodeMonitor(nc.logger, nc.ds, nodeName, nc.enqueueNodeForMonitor)
+	if err != nil {
+		return nil, err
+	}
+
+	nc.monitor = monitor
+
+	return monitor, nil
+}
+
+func (nc *NodeController) enqueueNodeForMonitor(key string) {
+	nc.queue.Add(key)
+}
+
+func (nc *NodeController) syncOrphans(node *longhorn.Node, collectedDataInfo map[string]*monitor.CollectedDiskInfo) error {
+	for diskName, diskInfo := range collectedDataInfo {
+		if err := nc.createOrphans(node, diskName, diskInfo); err != nil {
+			return errors.Wrapf(err, "failed to create orphans for disk %v", diskName)
+		}
+		if err := nc.deleteOrphans(node, diskName, diskInfo); err != nil {
+			return errors.Wrapf(err, "failed to delete orphans for disk %v", diskName)
+		}
+	}
+
+	return nil
+}
+
+func (nc *NodeController) deleteOrphans(node *longhorn.Node, diskName string, diskInfo *monitor.CollectedDiskInfo) error {
+	autoDeletionEnabled, err := nc.ds.GetSettingAsBool(types.SettingNameOrphanAutoDeletion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get %v setting", types.SettingNameOrphanAutoDeletion)
+	}
+
+	orphans, err := nc.ds.ListOrphansRO()
+	if err != nil {
+		return errors.Wrap(err, "unable to list orphans")
+	}
+
+	for _, orphan := range orphans {
+		if orphan.Status.OwnerID != nc.controllerID {
+			continue
+		}
+
+		switch {
+		case orphan.Spec.Type == longhorn.OrphanTypeReplica:
+			// Alignment between the orphan CRs and the collected orphaned directory names list.
+			// If the on-disk data is missing, e.g. deleted by users by manual, we need to delete the orphaned CR.
+			// Orphan controller won't stat/check the on-disk data and isn't aware of the missing replica directory,
+			// If the on-disk data is missing, i.e. not found in the orphaned directory names list,
+			// we need to delete the orphaned CR.
+			if onDiskDataMissing := nc.isOnDiskDataMissing(orphan, diskInfo); onDiskDataMissing {
+				if err := nc.ds.DeleteOrphan(orphan.Name); err != nil && !apierrors.IsNotFound(err) {
+					return errors.Wrapf(err, "failed to delete orphan %v", orphan.Name)
+				}
+				diskInfo.OrphanedReplicaDirectoryNames[orphan.Spec.Parameters[longhorn.OrphanDataName]] = longhorn.OrphanCRStatusDeleted
+				continue
+			}
+		}
+
+		dataCleanableCondition := types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeDataCleanable)
+		if dataCleanableCondition.Status == longhorn.ConditionStatusUnknown {
+			continue
+		}
+
+		if autoDeletionEnabled || dataCleanableCondition.Status == longhorn.ConditionStatusFalse {
```

Comment:
Should it be `if autoDeletionEnabled || dataCleanableCondition.Status == longhorn.ConditionStatusTrue`?

Reply:
When orphan controller detects a node is evicted, disk is evicted or something happened, it means the data is not accessible and not cleanable anymore. Thus `dataCleanableCondition.Status` of the orphan CR is set to  `longhorn.ConditionStatusFalse`.  

Then, this logic is to clean up the orphan CR that their data is not accessible, so check `dataCleanableCondition.Status == longhorn.ConditionStatusTrue` here.

Reply:
Thanks for the explanation!